pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -298,63 +350,75 @@ func (e *DockercfgController) syncServiceAccount(key string) error {
 	return err
 }
 
-const (
-	tokenSecretWaitInterval = 20 * time.Millisecond
-	tokenSecretWaitTimes    = 100
-)
+// getTokenSecret return a token secret for a given service account created by this
+// controller or returns nil when there is no token secret created yet.
+func (e *DockercfgController) getTokenSecret(serviceAccount *api.ServiceAccount) *api.Secret {
+	var tokenSecret *api.Secret
+	for _, s := range e.secretsCache.ListKeys() {
+		if strings.HasPrefix(s, serviceAccount.Namespace+"/"+osautil.GetTokenSecretNamePrefix(serviceAccount)) {
+			uncastSecret, _, _ := e.secretsCache.GetByKey(s)
+			if secret := uncastSecret.(*api.Secret); len(secret.Annotations) > 0 {
+				if createdBy, exists := secret.Annotations[api.CreatedByAnnotation]; exists && createdBy == "create_dockercfg_secrets" {
+					tokenSecret = secret
+					break
+				}
+			}
+		}
+	}
+	return tokenSecret
+}
 
 // createTokenSecret creates a token secret for a given service account.  Returns the name of the token
-func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, error) {
-	tokenSecret := &api.Secret{
+func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, bool, error) {
+	tokenSecret := e.getTokenSecret(serviceAccount)
+	if tokenSecret != nil {
+		return tokenSecret, len(tokenSecret.Data[api.ServiceAccountTokenKey]) > 0, nil
+	}
+	tokenSecret = &api.Secret{
 		ObjectMeta: api.ObjectMeta{
 			Name:      secret.Strategy.GenerateName(osautil.GetTokenSecretNamePrefix(serviceAccount)),
 			Namespace: serviceAccount.Namespace,
 			Annotations: map[string]string{
 				api.ServiceAccountNameKey: serviceAccount.Name,
 				api.ServiceAccountUIDKey:  string(serviceAccount.UID),
+				api.CreatedByAnnotation:   "create_dockercfg_secrets",
```

Comment:
(i pushed recent version where this is a `const` and it is also used to filter just tokens we created here instead of watching all secret tokens)